homeassistant/components/tplink/select.py

```diff
@@ -0,0 +1,89 @@
+"""Support for TPLink select entities."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Final, cast
+
+from kasa import Device, Feature
+
+from homeassistant.components.select import SelectEntity, SelectEntityDescription
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import TPLinkConfigEntry
+from .entity import (
+    CoordinatedTPLinkFeatureEntity,
+    TPLinkDataUpdateCoordinator,
+    TPLinkFeatureEntityDescription,
+    async_refresh_after,
+)
+
+
+@dataclass(frozen=True, kw_only=True)
+class TPLinkSelectEntityDescription(
+    SelectEntityDescription, TPLinkFeatureEntityDescription
+):
+    """Base class for a TPLink feature based sensor entity description."""
+
+
+SELECT_DESCRIPTIONS: Final = [
+    TPLinkSelectEntityDescription(
+        key="light_preset",
+    ),
+]
+
+SELECT_DESCRIPTIONS_MAP = {desc.key: desc for desc in SELECT_DESCRIPTIONS}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: TPLinkConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up sensors."""
+    data = config_entry.runtime_data
+    parent_coordinator = data.parent_coordinator
+    children_coordinators = data.children_coordinators
+    device = parent_coordinator.device
+
+    entities = CoordinatedTPLinkFeatureEntity.entities_for_device_and_its_children(
+        device=device,
+        coordinator=parent_coordinator,
+        feature_type=Feature.Type.Choice,
+        entity_class=Select,
+        descriptions=SELECT_DESCRIPTIONS_MAP,
+        child_coordinators=children_coordinators,
+    )
+    async_add_entities(entities)
+
+
+class Select(CoordinatedTPLinkFeatureEntity, SelectEntity):
+    """Representation of a tplink select entity."""
+
+    entity_description: TPLinkSelectEntityDescription
+
+    def __init__(
+        self,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        *,
+        feature: Feature,
+        description: TPLinkFeatureEntityDescription,
+        parent: Device | None = None,
+    ) -> None:
+        """Initialize a select."""
+        super().__init__(
+            device, coordinator, feature=feature, description=description, parent=parent
+        )
+        self._attr_options = cast(list, self._feature.choices)
```

Comment:
Not critical for this pr but this should probably go in `_async_update_attrs` and we can drop the constructor.  Although the library doesn't currently support indirection for `choices` it should in the future (same for `units`) so it can handle updates to the underlying choices.

Reply:
We will revisit this when we are that far. I haven't looked at the internal code, but I'd guess that it will be more performant when it's not being set on every update cycle :+1: